MatrixSDK/VoIP/CallKit/MXCallKitAdapter.m

```diff
@@ -0,0 +1,285 @@
+    update.remoteHandle = handle;
+    update.localizedCallerName = caller.displayname;
+    update.hasVideo = call.isVideoCall;
+    update.supportsHolding = NO;
+    update.supportsGrouping = NO;
+    update.supportsUngrouping = NO;
+    update.supportsDTMF = NO;
+    
+    [self.provider reportNewIncomingCallWithUUID:callUUID update:update completion:^(NSError * _Nullable error) {
+        if (error)
+        {
+            [call hangup];
+            return;
+        }
+        
+        self.calls[callUUID] = call;
+        
+#ifdef MX_CALL_STACK_JINGLE
+        // Workaround from https://forums.developer.apple.com/message/169511 suggests configuring audio in the
+        // completion block of the `reportNewIncomingCallWithUUID:update:completion:` method instead of in
+        // `provider:performAnswerCallAction:` per the WWDC examples.
+        [MXJingleCallAudioSessionConfigurator configureAudioSessionForVideoCall:call.isVideoCall];
+#endif
+    }];
+    
+}
+
+- (void)reportCall:(MXCall *)call startedConnectingAtDate:(nullable NSDate *)date
+{
+    [self.provider reportOutgoingCallWithUUID:call.callUUID startedConnectingAtDate:date];
+}
+
+- (void)reportCall:(MXCall *)call connectedAtDate:(nullable NSDate *)date
+{
+    [self.provider reportOutgoingCallWithUUID:call.callUUID connectedAtDate:date];
+}
+
++ (BOOL)callKitAvailable
+{
+    return [NSProcessInfo.processInfo isOperatingSystemAtLeastVersion:(NSOperatingSystemVersion){10,0,0}];
+}
+
+#pragma mark - CXProviderDelegate
+
+- (void)providerDidReset:(CXProvider *)provider
+{
+    NSLog(@"Provider did reset");
+}
+
+- (void)provider:(CXProvider *)provider didActivateAudioSession:(AVAudioSession *)audioSession
+{
+#ifdef MX_CALL_STACK_JINGLE
+    [MXJingleCallAudioSessionConfigurator audioSessionDidActivate:audioSession];
+#endif
+}
+
+- (void)provider:(CXProvider *)provider didDeactivateAudioSession:(AVAudioSession *)audioSession
+{
+#ifdef MX_CALL_STACK_JINGLE
+    [MXJingleCallAudioSessionConfigurator audioSessionDidDeactivate:audioSession];
+#endif
+}
+
+- (void)provider:(CXProvider *)provider performStartCallAction:(CXStartCallAction *)action
+{
+    MXCall *call = self.calls[action.callUUID];
+    if (!call)
+    {
+        [action fail];
+        return;
+    }
+    
+#ifdef MX_CALL_STACK_JINGLE
+    [MXJingleCallAudioSessionConfigurator configureAudioSessionForVideoCall:call.isVideoCall];
+#endif
+    
+    [action fulfill];
+}
+
+- (void)provider:(CXProvider *)provider performAnswerCallAction:(CXAnswerCallAction *)action
+{
+    MXCall *call = self.calls[action.callUUID];
+    if (!call)
+    {
+        [action fail];
+        return;
+    }
+    
+    [call answer];
+    
+    [action fulfill];
+}
+
+- (void)provider:(CXProvider *)provider performEndCallAction:(CXEndCallAction *)action
+{
+    MXCall *call = self.calls[action.callUUID];
+    if (!call)
+    {
+        [action fail];
+        return;
+    }
+    
+    [call hangup];
+    self.calls[action.UUID] = nil;
```

Comment:
I would prefer to use removeObjectForKey. I always fear about exception thrown because of setting nil value in NSMutableDictionary.